lib/mapping/CCampaignHandler.cpp

```diff
@@ -151,6 +180,257 @@ std::string CCampaignHandler::readLocalizedString(CBinaryReader & reader, std::s
 	return VLC->generaltexth->translate(stringID.get());
 }
 
+CCampaignHeader CCampaignHandler::readHeaderFromJson(JsonNode & reader, std::string filename, std::string modName, std::string encoding )
+{
+	CCampaignHeader ret;
+
+	ret.version = reader["version"].Integer();
+	if(ret.version < CampaignVersion::VCMI_MIN || ret.version > CampaignVersion::VCMI_MAX)
+	{
+		logGlobal->info("VCMP Loading: Unsupported campaign %s version %d", filename, ret.version);
+		return ret;
+	}
+	
+	ret.version = CampaignVersion::VCMI;
+	ret.mapVersion = reader["campaignId"].Integer();
+	ret.name = reader["name"].String();
+	ret.description = reader["description"].String();
+	ret.difficultyChoosenByPlayer = reader["allowDifficultySelection"].Bool();
+	//skip ret.music because it's unused in vcmi
+	ret.filename = filename;
+	ret.modName = modName;
+	ret.encoding = encoding;
+	ret.valid = true;
+	return ret;
+}
+
+CCampaignScenario CCampaignHandler::readScenarioFromJson(JsonNode & reader, std::string filename, std::string modName, std::string encoding, int version, int mapVersion)
+{
+	auto prologEpilogReader = [](JsonNode & identifier) -> CCampaignScenario::SScenarioPrologEpilog
+	{
+		CCampaignScenario::SScenarioPrologEpilog ret;
+		ret.hasPrologEpilog = !identifier.isNull();
+		if(ret.hasPrologEpilog)
+		{
+			ret.prologVideo = identifier["video"].Integer();
+			ret.prologMusic = identifier["music"].Integer();
+			ret.prologText = identifier["text"].String();
+		}
+		return ret;
+	};
+
+	CCampaignScenario ret;
+	ret.conquered = false;
+	ret.mapName = reader["map"].String();
+	for(auto & g : reader["precoditions"].Vector())
+		ret.preconditionRegions.insert(g.Integer());
+
+	ret.regionColor = reader["color"].Integer();
+	ret.difficulty = reader["difficulty"].Integer();
+	ret.regionText = reader["regionText"].String();
+	ret.prolog = prologEpilogReader(reader["prolog"]);
+	ret.epilog = prologEpilogReader(reader["epilog"]);
+
+	ret.travelOptions = readScenarioTravelFromJson(reader, version);
+
+	return ret;
+}
+
+CScenarioTravel CCampaignHandler::readScenarioTravelFromJson(JsonNode & reader, int version )
+{
+	CScenarioTravel ret;
+
+	std::map<std::string, ui8> heroKeepsMap = {
+		{"experience", 1},
+		{"primarySkill", 2},
+		{"secondarySkill", 4},
+		{"spells", 8},
+		{"artifacts", 16}

```

Comment:
Do these constants have enumeration somewhere? If not - they should IMO

(same for other maps below)

Reply:
enumeration for strings?
All those maps are temporary to transform data to internal legacy h3c format.
Take a look at campToERMU methods - h3c has very specific format (bit masks, etc)

Reply:
Not for strings. But for bit masks from H3, e.g.
```c++
enum EHeroKeeps
{
    EXPERIENCE = 1,
    PRIMARY_SKILLS = 2,
    ....
}
```
Probably needs to be wrapped in namespace to avoid leaking into global namespace - since proper enum class won't be suitable here.

Reply:
Again, we won't use them later. I don't think we need enums which are used in only one place